File: cpp_codes/2023/december/17-Max Sum without Adjacents.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
//User function template for C++
class Solution{
public:	
	// calculate the maximum sum with out adjacent
	int findMaxSum(int *arr, int n) {
	    // code here
	    if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return arr[0];
        }

        // Initialize an array to store the maximum sum up to each index
        int dp[n];

        // Base cases
        dp[0] = arr[0];
        dp[1] = max(arr[0], arr[1]);

        // Fill the dp array using the recurrence relation
        for (int i = 2; i < n; i++) {
            dp[i] = max(dp[i - 1], dp[i - 2] + arr[i]);
        }

        // Return the maximum sum
        return dp[n - 1];
	}
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        int arr[n];
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.findMaxSum(arr, n);
        cout << ans << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: java_codes/2023/december/17-Max Sum without Adjacents.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());
        while (tc-- > 0) {
            String[] inputLine;
            int n = Integer.parseInt(br.readLine().trim());
            int[] arr = new int[n];
            inputLine = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(inputLine[i]);
            }

            int ans = new Solution().findMaxSum(arr, n);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends

//User function Template for Java
class Solution {
    int findMaxSum(int arr[], int n) {
        if(n == 1){
            return arr[0];
        }
        
        int dp[] = new int[n];
        dp[0] = arr[0];
        dp[1] = Math.max(arr[0], arr[1]);
        
        for(int i = 2; i<n; i++){
            dp[i] = Math.max(dp[i-1], arr[i] + dp[i-2]);
        }
        
        return dp[n-1];
    }
}